tikv-client/src/main/java/com/pingcap/tikv/KVClient.java

```diff
@@ -0,0 +1,227 @@
+    Iterator<Kvrpcpb.KvPair> iterator = scanIterator(conf, clientBuilder, startKey, limit);
+    List<Kvrpcpb.KvPair> result = new ArrayList<>();
+    iterator.forEachRemaining(result::add);
+    return result;
+  }
+
+  /** A Batch containing the region, a list of keys to send */
+  private static final class Batch {
+    private final TiRegion region;
+    private final List<ByteString> keys;
+
+    Batch(TiRegion region, List<ByteString> keys) {
+      this.region = region;
+      this.keys = keys;
+    }
+  }
+
+  /**
+   * Append batch to list and split them according to batch limit
+   *
+   * @param batches a grouped batch
+   * @param region region
+   * @param keys keys
+   * @param limit batch max limit
+   */
+  private void appendBatches(
+      List<Batch> batches, TiRegion region, List<ByteString> keys, int limit) {
+    List<ByteString> tmpKeys = new ArrayList<>();
+    for (int i = 0; i < keys.size(); i++) {
+      if (i >= limit) {
+        batches.add(new Batch(region, tmpKeys));
+        tmpKeys.clear();
+      }
+      tmpKeys.add(keys.get(i));
+    }
+    if (!tmpKeys.isEmpty()) {
+      batches.add(new Batch(region, tmpKeys));
+    }
+  }
+
+  /**
+   * Group by list of keys according to its region
+   *
+   * @param keys keys
+   * @return a mapping of keys and their region
+   */
+  private Map<TiRegion, List<ByteString>> groupKeysByRegion(Set<ByteString> keys) {
+    return keys.stream()
+        .collect(Collectors.groupingBy(clientBuilder.getRegionManager()::getRegionByKey));
+  }
+
+  /**
+   * Send batchPut request concurrently
+   *
+   * @param backOffer current backOffer
+   * @param batches list of batch to send
+   */
+  private List<KvPair> sendBatchGet(BackOffer backOffer, List<Batch> batches, long version) {
+    for (Batch batch : batches) {
+      completionService.submit(
+          () -> {
+            RegionStoreClient client = clientBuilder.build(batch.region);
+            BackOffer singleBatchBackOffer = ConcreteBackOffer.create(backOffer);
+            List<ByteString> keys = batch.keys;
+            try {
+              return client.batchGet(singleBatchBackOffer, keys, version);
+            } catch (final TiKVException e) {
+              // TODO: any elegant way to re-split the ranges if fails?
+              singleBatchBackOffer.doBackOff(BackOffFunction.BackOffFuncType.BoRegionMiss, e);
+              logger.warn("ReSplitting ranges for BatchPutRequest");
+              // recursive calls
+              return batchGet(singleBatchBackOffer, batch.keys, version);
+            }
+          });
+    }
+    try {
+      List<KvPair> result = new ArrayList<>();
+      for (int i = 0; i < batches.size(); i++) {
+        result.addAll(
+            completionService.take().get(BackOffer.BATCH_GET_MAX_BACKOFF, TimeUnit.SECONDS));
+      }
+      return result;
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```

Comment:
I do not understand why we need interrupt thread here.

Reply:
do you have any good ideas to deal with InterruptedException?

Reply:
```
try {
  Thread.sleep(100);
} catch (InterruptedException ex) {
  throw new RuntimeException(ex);
}
```
Above is what I found on stackoverflow

Reply:
Well, in case you do not want to do anything to interrupt exception, I think it is better to restore the interrupted status. What do you think, @marsishandsome ?